pkg/store/labelpb/label.go

```diff
@@ -346,6 +352,44 @@ func HashWithPrefix(prefix string, lbls []ZLabel) uint64 {
 	return xxhash.Sum64(b)
 }
 
+// ValidateLabels validates label names and values (checks for empty
+// names and values, out of order labels and duplicate label names)
+// Returns appropriate error if validation fails on a label.
+func ValidateLabels(lbls []ZLabel) error {
```

Comment:
Any concerns with increasing write latency due to this added validation?

Reply:
We could benchmark it so that we can make an informed decision about how much the latency changes. Maybe the validation is negligible compared to the storing or the metrics in the TSDB, maybe not.

Reply:
Benchmarking if possible would be great. If that's too complicated, I can keep an eye on this after deploying.

Now that you mentioned TSDB, I wonder why it even accepts writing samples with OOO labels. Should we also open an issue in prometheus/prometheus?

Reply:
I think that is a valid concern, I'd see how complicated it would be to add a benchmark for this.

As for TSDB, I'm not sure whether this was ever a requirement or a problem, I've seen various reference to ordering sprinkled throughout the repo / codebase (https://github.com/prometheus/prometheus/issues/8861#issuecomment-855885421, https://github.com/prometheus/prometheus/pull/10532, https://github.com/prometheus/prometheus/pull/5372), but I cannot find any authoritative doc saying how labels must be ordered at time of ingestion.

Reply:
I actually realized that we can simplify the validation, since we are expecting the labels to be ordered, we can first check order and then duplication, meaning we don't have to use a map to check duplicates (saving us allocations).

I ran benchmarks from https://github.com/thanos-io/thanos/pull/5533 on old and new code and there's no discernible difference:
```
name                                     old time/op    new time/op    delta
WriterTimeSeriesWithSingleLabel_10-12      4.08µs ± 0%    4.02µs ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWithSingleLabel_100-12     22.6µs ± 0%    21.3µs ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWithSingleLabel_1000-12     238µs ± 0%     197µs ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWith10Labels_10-12         6.18µs ± 0%    5.57µs ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWith10Labels_100-12        37.9µs ± 0%    41.1µs ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWith10Labels_1000-12        368µs ± 0%     430µs ± 0%   ~     (p=1.000 n=1+1)

name                                     old alloc/op   new alloc/op   delta
WriterTimeSeriesWithSingleLabel_10-12        411B ± 0%      411B ± 0%   ~     (all equal)
WriterTimeSeriesWithSingleLabel_100-12       413B ± 0%      411B ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWithSingleLabel_1000-12      454B ± 0%      453B ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWith10Labels_10-12           410B ± 0%      410B ± 0%   ~     (all equal)
WriterTimeSeriesWith10Labels_100-12          411B ± 0%      412B ± 0%   ~     (p=1.000 n=1+1)
WriterTimeSeriesWith10Labels_1000-12         460B ± 0%      471B ± 0%   ~     (p=1.000 n=1+1)

name                                     old allocs/op  new allocs/op  delta
WriterTimeSeriesWithSingleLabel_10-12        9.00 ± 0%      9.00 ± 0%   ~     (all equal)
WriterTimeSeriesWithSingleLabel_100-12       9.00 ± 0%      9.00 ± 0%   ~     (all equal)
WriterTimeSeriesWithSingleLabel_1000-12      9.00 ± 0%      9.00 ± 0%   ~     (all equal)
WriterTimeSeriesWith10Labels_10-12           9.00 ± 0%      9.00 ± 0%   ~     (all equal)
WriterTimeSeriesWith10Labels_100-12          9.00 ± 0%      9.00 ± 0%   ~     (all equal)
WriterTimeSeriesWith10Labels_1000-12         9.00 ± 0%      9.00 ± 0%   ~     (all equal)
```